### ISSUE TITLE
Questions on invoking quantization aware training from full-precision checkpoints

### ISSUE THREAD

**Issue by lehahoang** [NervanaSystems/distiller#408]
Hi,

In this post, I would like to raise a few questions:
1. How can we justify if quantization-aware training from full-precision checkpoints gives us the correct result?
2. How do we capture quantized weights during the quantization-aware training process?
3. Are weights quantized on-the-fly by using the module quantize_wts in this [tutorial](https://nervanasystems.github.io/distiller/quantization.html) without being temporarily saved?
Here is my experimental setup:
- A full-precision VGG-16 on CIFAR-10 dataset with 88.81% top-1 accuracy
- A quantization.yaml file:
```
quantizers:
  linear_quantizer:
    class: QuantAwareTrainRangeLinearQuantizer
    bits_activations: 8
    bits_weights: 8
    mode: 'SYMMETRIC'  # Can try "SYMMETRIC" as well
    per_channel_wts: True

policies:
    - quantizer:
        instance_name: linear_quantizer
      starting_epoch: 0
      ending_epoch: 100
      frequency: 1
```
- 5000 samples of validation dataset
- My command line, which is inspired from [here](https://github.com/NervanaSystems/distiller/pull/309), is:

`CUDA_VISIBLE_DEVICES=1 python compress_classifier.py -a vgg16_cifar --lr 0.001 ./data.cifar --epochs 91 --compress=vgg16/quant.yaml --wd=0.0001 --vs=0.1 --resume-from vgg16/checkpoint.pth.tar --deterministic `

My assumption in post-training quantization is that weights are quantized by using wrapper_module then are forwarded to the network. This is not an issue of capturing quantized weights during the process. Regarding quantization-aware training, weights are quantized magically "somewhere" in the process. Therefore, I am not sure if my command line invoked weights to be quantized. Please correct me if I am wrong. 

Thank you.

**Comment by lehahoang** [comment]
Hi guys,
I have recently found out that there is a comment in the range_linear.py might gives the hint to my third question:
```

# We don't quantize the learned weights of embedding layers per-channel, because they're used
# as inputs in subsequent layers and we don't support per-channel activations quantization yet
```

What I perceived from reading this comment is that weights of the full-precision model are not truly quantized. As a result, the dot products between input activations and weights are still represented in Float32 for each layer. Only the output activations are quantized to lower bitwidth using FakeLinearQuantization class. And this is called "simulated quantization", which is described in [here](https://nervanasystems.github.io/distiller/quantization.html).

Please correct me if I am wrong.
Thank you.

**Comment by guyjacob** [comment]
Hi @lehahoang,

Apologies for the very delayed response.

In case this is still relevant, please take a look here:
https://nervanasystems.github.io/distiller/design.html
Specifically, the "**Quantization-Aware Training**" section near the bottom.
I think the text there should explain the "magic" that happens during quantization-aware training.
A couple of extra points:
* What's missing from that explanation is how weights quantization is actually invoked during training. This is done by calling the `Quantizer`'s `quantize_params()` function. When specifying a quantization policy, as you do in the YAML file above, the policy takes care of this for you. See [here](https://github.com/NervanaSystems/distiller/blob/master/distiller/policy.py#L317).
* At all times during training you can inspect the modules and see both the float and quantized versions of the weights via the `float_weight` and `weight` parameters, respectively (this applies to the saved checkpoints as well).
* Usually in quant-aware training, the "quantized" weights will still be FP32 values - but they will be discretized. That is - for 8-bit quantization, there will be 256 different FP32 values. That's the "simulated quantization" you correctly referred to above.

Regarding the comment in the code you mentioned - it actually refers to a fairly limited context - embedding layers during **post-training** quantization only. It also does not mean the weights of the embeddings are not quantized at all, just that we don't support per-channel quantization for embeddings. In any case, it's not related to quantization-aware training.

Again, sorry for the delayed response. Let us know if you have more questions.

**Comment by lehahoang** [comment]
Hi @guyjacob ,

Thank you for your kind reply. It helps me a lot actually.
I want to add more points that in case you want to apply the training policy in YAML file, please include "--reset-optimizer" to the bash command, otherwise, the training policy of the pre-trained model will be used instead. Could you please add it into the example in quantization folder so that it could help new user?

I think I will close the issue now.
Best regards,
Le-Ha Hoang

**Comment by killerzula** [comment]
> - At all times during training you can inspect the modules and see both the float and quantized versions of the weights via the float_weight and weight parameters, respectively (this applies to the saved checkpoints as well).

Thanks for your help @guyjacob. Does it mean that I can export my model after QAT by just creating a state_dict and storing the `.weights` inside of it?
